### ISSUE TITLE
Pyright ignores type equality check

### ISSUE THREAD

**Issue by joe733** [microsoft/pyright#4633]
**Describe the bug**
Variables with union of types gets combination of types in comparisons, which is reasonable. But if `val_1: int | str = ""`, `val_2: int | str = ""` and I check `if type(val_1) == type(val_2)` then `return val_1 >= val_2`. Pyright complains:

```
Operator ">=" not supported for types "int | str" and "int | str"
Operator ">=" not supported for types "int" and "str"
Operator ">=" not supported for types "str" and "int"`
```

**To Reproduce**
Here's another code snippet:
```py
from datetime import datetime
from typing import Union

def between(
    value: Union[int, float, str, datetime],
    min_val: Union[int, float, str, datetime] = 0,
    max_val: Union[int, float, str, datetime] = 0,
) -> bool:
    if type(value) != type(min_val) != type(max_val):
        raise AssertionError("`min` and `max` must be of same type")
    return min_val <= value and max_val >= value
```

**Expected behavior**
Pyright understand that the comparison in the return statement is done only between same type.

**Screenshots or Code**
![image](https://user-images.githubusercontent.com/38415384/218776091-7d70669d-58e1-4e4f-b6be-85877feda620.png)

**VS Code extension or command-line**
VSCode Extension: Pylance v2023.2.20

**Additional context**
Or is there another way to do it?

**Comment by erictraut** [comment]
Pyright is working as designed here. For more information about how type narrowing and type guards work in a static type checker, please refer to [this documentation](https://github.com/microsoft/pyright/blob/main/docs/type-concepts.md#type-narrowing).

There are a couple of issues here:
1. The expression form that you're using here is not a supported type guard pattern. See [this list](https://github.com/microsoft/pyright/blob/main/docs/type-concepts.md#type-guards) for supported patterns.
2. The type narrowing process narrows the type of individual expressions. It does not handle conditional dependencies between expressions such as "`x` has type `int` if `y` also has type `int` otherwise `x` is type `str`". Tracking such dependencies is not computationally feasible, and I'm not aware of any static type checker or compiler in any language that does this.

**Comment by erictraut** [comment]
One option is to use a constrained TypeVar. This doesn't work with argument default values though, so you wouldn't be able to use `= 0` for some parameters.

```python
T = TypeVar("T", int, float, str, datetime)


def between(value: T, min_val: T, max_val: T) -> bool:
    return min_val <= value and max_val >= value
```

**Comment by joe733** [comment]
That helps, thank you!
